test/interpreter/function-substitute.spec.ts

```diff
@@ -0,0 +1,74 @@
+import {ErrorType, HyperFormula} from '../../src'
+import {adr, detailedError} from '../testUtils'
+import {ErrorMessage} from '../../src/error-message'
+
+describe('Function SUBSTITUTE', () => {
+  it('should take three or four parameters', () => {
+    const engine = HyperFormula.buildFromArray([
+      ['=SUBSTITUTE("foo", "f")'],
+      ['=SUBSTITUTE("foobar", "o", "uu", 4, 5)'],
+    ])
+
+    expect(engine.getCellValue(adr('A1'))).toEqual(detailedError(ErrorType.NA, ErrorMessage.WrongArgNumber))
+    expect(engine.getCellValue(adr('A2'))).toEqual(detailedError(ErrorType.NA, ErrorMessage.WrongArgNumber))
+  })
+
+  it('should substitute new text for all occurrences of old text in a string', () => {
+    const engine = HyperFormula.buildFromArray([
+      ['=SUBSTITUTE("foo", "f", "bb")'],
+      ['=SUBSTITUTE("foobar", "o", "uu")'],
+    ])
+
+    expect(engine.getCellValue(adr('A1'))).toEqual('bboo')
+    expect(engine.getCellValue(adr('A2'))).toEqual('fuuuubar')
+  })
+
+  it('should substitute new text for nth occurrence of a string', () => {
+    const engine = HyperFormula.buildFromArray([
+      ['=SUBSTITUTE("foobar", "o", "f", 1)'],
+      ['=SUBSTITUTE("foobar", "o", "OO", 2)'],
+      ['=SUBSTITUTE("foobar", "o", "OO", 3)'],
+      ['=SUBSTITUTE("fofofofofo", "o", "u", 4)']
+    ])
+
+    expect(engine.getCellValue(adr('A1'))).toEqual('ffobar')
+    expect(engine.getCellValue(adr('A2'))).toEqual('foOObar')
+    expect(engine.getCellValue(adr('A3'))).toEqual('foobar')
+    expect(engine.getCellValue(adr('A4'))).toEqual('fofofofufo')
+  })
+
```

Comment:
can we have a test when pattern is self-overlapping?

SUBSTITUTE("fooobar", "oo", "x")

Reply:
done